WKWebView sometimes allows app links to open when reloading after a web process crash

davidquesada opened the pull request:
#### 075cc18c58a3ebdd1074093a5dae9dc483cf6b91
<pre>
WKWebView sometimes allows app links to open when reloading after a web process crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=288159">https://bugs.webkit.org/show_bug.cgi?id=288159</a>
<a href="https://rdar.apple.com/141515904">rdar://141515904</a>

Reviewed by Alex Christensen.

In some cases, reloading the web view after a web process crash can result in the
WKWebView opening the current URL as an app link and punching out to the app, which
can be quite annoying for the user. Ordinarily, web view reloads arent&apos;t be eligible
for opening app links because doing so requires that the navigation is from one host
to a different one. But when performing a reload after a web process crash, the main
frame&apos;s URL has been reset to null after the process termination, and having a null
URL technically satisifies the different-host requirement because the previously
loaded URL&apos;s host is *anything except null*.

Prevent these unexpected app launches by adding a requirement that `shouldOpenAppLinks`
can only be true when the main frame&apos;s URL is non-null, since opening app links
automatically only makes sense when there is actual content loaded in the web view
that the user might have explicitly interacted with. An exception to this requirement
is made if the web view has never committed any provisional navigations, as is the
case when one web view opens another to a URL that ends up resolving to an app link.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
    Implement the aforementioned check.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ShouldOpenAppLinks.mm: Added.
(-[ShouldOpenAppLinksTestNavigationDelegate init]):
    Set up a navigation delegate for the tests that retains the most recent
    WKNavigationAction and reloads the web view on web process crash. Manually
    performing this reload is necessary because the WebPageProxy doesn&apos;t internally
    do this reload itself if the crash was requested by the client.
(shouldOpenAppLinksTestServer):
(TEST(ShouldOpenAppLinks, DisallowAppLinksWhenReloadingAfterWebProcessCrash)):
(TEST(ShouldOpenAppLinks, DisallowAppLinksWhenReloadingAfterWebProcessCrashAfterFollowingLink)):
    Verify that WKNavigationAction._shouldOpenAppLinks (which reflects the same
    underlying effective policy that WebKit internally uses to decide whether to
    try opening URLs as app links) is false when deciding the navigatiom policy
    for a reload being performed after a web process crash. Add a distinct test
    case for the case where the web view has been navigated by the user clicking
    a link. In that case, unlike the first, it appears the user-initiated nature
    of the navigation to a second page attaches a policy of
    ShouldOpenExternalURLsPolicy::ShouldAllow to the HistoryItem being reloaded,
    which was previously allowing `shouldOpenAppLinks` to evaluate to true.

Canonical link: <a href="https://commits.webkit.org/290885@main">https://commits.webkit.org/290885@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c9e86d9c8a562a3d60adcd0be6b2836f24ca745

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91283 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10818 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/313 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96266 "") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42005 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93333 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11203 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19230 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/96266 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27629 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94284 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8544 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82680 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/96266 "") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8314 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/261 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41148 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78635 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/271 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98257 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18458 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79129 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18714 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78513 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78331 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19387 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22843 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/196 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11627 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18457 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23757 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18176 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21637 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19948 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ShouldOpenAppLinks.mm:
```diff
@@ -0,0 +1,138 @@
+
+@end
+
+static TestWebKitAPI::HTTPServer testServer(void)
```
(void) should be ()
Let's give this a slightly more specific name, like appLinkTestServer.  With unified builds this name is likely to collide.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ShouldOpenAppLinks.mm:
```diff
@@ -0,0 +1,138 @@
+    auto delegate = adoptNS([[ShouldOpenAppLinksTestNavigationDelegate alloc] init]);
+    [webView setNavigationDelegate:delegate.get()];
+
+    auto &&server = testServer();
```
WebKit style prefers the space after the &&, but in this case you should omit the &&.  Same below.

davidquesada replied:
I figured I was doing something wrong without this, since I was getting some build errors I couldn't quite figure out:
```
In file included from /Volumes/Volatile/DerivedData/Build/Products/Release-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource27-mm.mm:1:
In file included from /Volumes/Source/OpenSource/Tools/TestWebKitAPI/TestWebKitAPIPrefix.h:57:
In file included from /Volumes/Volatile/DerivedData/Build/Products/Release-iphonesimulator/usr/local/include/wtf/text/StringView.h:31:
In file included from /Volumes/Volatile/DerivedData/Build/Products/Release-iphonesimulator/usr/local/include/wtf/RetainPtr.h:29:
In file included from /Volumes/Volatile/DerivedData/Build/Products/Release-iphonesimulator/usr/local/include/wtf/HashTraits.h:26:
In file included from /Volumes/Volatile/DerivedData/Build/Products/Release-iphonesimulator/usr/local/include/wtf/HashFunctions.h:26:
In file included from /Volumes/Volatile/DerivedData/Build/Products/Release-iphonesimulator/usr/local/include/wtf/RefPtr.h:28:
/Volumes/Volatile/DerivedData/Build/Products/Release-iphonesimulator/usr/local/include/wtf/Ref.h:55:12: error: member access into incomplete type 'TestWebKitAPI::HTTPServer::RequestData'
   55 |         ref.ref();
      |            ^
/Volumes/Volatile/DerivedData/Build/Products/Release-iphonesimulator/usr/local/include/wtf/Ref.h:91:34: note: in instantiation of member function 'WTF::DefaultRefDerefTraits<TestWebKitAPI::HTTPServer::RequestData>::ref' requested here
   91 |         : m_ptr(&RefDerefTraits::ref(other.get()))
      |                                  ^
In file included from /Volumes/Volatile/DerivedData/Build/Products/Release-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource27-mm.mm:1:
In file included from /Volumes/Source/OpenSource/Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:29:
/Volumes/Source/OpenSource/Tools/TestWebKitAPI/cocoa/HTTPServer.h:43:7: note: in instantiation of member function 'WTF::Ref<TestWebKitAPI::HTTPServer::RequestData>::Ref' requested here
   43 | class HTTPServer {
      |       ^
/Volumes/Source/OpenSource/Tools/TestWebKitAPI/cocoa/HTTPServer.h:46:12: note: forward declaration of 'TestWebKitAPI::HTTPServer::RequestData'
   46 |     struct RequestData;
      |            ^
```

I'll take another look into this to figure what I'm missing that's stopping me from writing this in a way that makes more sense.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ShouldOpenAppLinks.mm:
```diff
@@ -0,0 +1,138 @@
+TEST(ShouldOpenAppLinks, DisallowAppLinksWhenReloadingAfterWebProcessCrash)
+{
+    auto webView = adoptNS([[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600)]);
+    auto delegate = adoptNS([[ShouldOpenAppLinksTestNavigationDelegate alloc] init]);
```
Nit: I like adoptNS([ClassName new]) for its brevity, but that's not important at all.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ShouldOpenAppLinks.mm:
```diff
@@ -0,0 +1,138 @@
+
+- (void)dealloc
+{
+    [_lastNavigationAction release];
```
Please use an ivar RetainPtr instead of all this.  That'll make our transition to ARC less difficult.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ShouldOpenAppLinks.mm:
```diff
@@ -0,0 +1,138 @@
+{
+    _hasNavigationFinished = false;
+    block();
+    TestWebKitAPI::Util::run(&_hasNavigationFinished);
```
_test_waitForDidFinishNavigation could replace a lot of this, but this is fine too.

davidquesada committed:
WKWebView sometimes allows app links to open when reloading after a web process crash
https://bugs.webkit.org/show_bug.cgi?id=288159
rdar://141515904

Reviewed by Alex Christensen.

In some cases, reloading the web view after a web process crash can result in the
WKWebView opening the current URL as an app link and punching out to the app, which
can be quite annoying for the user. Ordinarily, web view reloads arent't be eligible
for opening app links because doing so requires that the navigation is from one host
to a different one. But when performing a reload after a web process crash, the main
frame's URL has been reset to null after the process termination, and having a null
URL technically satisifies the different-host requirement because the previously
loaded URL's host is *anything except null*.

Prevent these unexpected app launches by adding a requirement that `shouldOpenAppLinks`
can only be true when the main frame's URL is non-null, since opening app links
automatically only makes sense when there is actual content loaded in the web view
that the user might have explicitly interacted with. An exception to this requirement
is made if the web view has never committed any provisional navigations, as is the
case when one web view opens another to a URL that ends up resolving to an app link.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
    Implement the aforementioned check.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ShouldOpenAppLinks.mm: Added.
(-[ShouldOpenAppLinksTestNavigationDelegate init]):
    Set up a navigation delegate for the tests that retains the most recent
    WKNavigationAction and reloads the web view on web process crash. Manually
    performing this reload is necessary because the WebPageProxy doesn't internally
    do this reload itself if the crash was requested by the client.
(shouldOpenAppLinksTestServer):
(TEST(ShouldOpenAppLinks, DisallowAppLinksWhenReloadingAfterWebProcessCrash)):
(TEST(ShouldOpenAppLinks, DisallowAppLinksWhenReloadingAfterWebProcessCrashAfterFollowingLink)):
    Verify that WKNavigationAction._shouldOpenAppLinks (which reflects the same
    underlying effective policy that WebKit internally uses to decide whether to
    try opening URLs as app links) is false when deciding the navigatiom policy
    for a reload being performed after a web process crash. Add a distinct test
    case for the case where the web view has been navigated by the user clicking
    a link. In that case, unlike the first, it appears the user-initiated nature
    of the navigation to a second page attaches a policy of
    ShouldOpenExternalURLsPolicy::ShouldAllow to the HistoryItem being reloaded,
    which was previously allowing `shouldOpenAppLinks` to evaluate to true.

Canonical link: https://commits.webkit.org/290885@main
